gk/main.c

```diff
@@ -548,6 +555,67 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_addr *nexthop_mac = arg;
+
+	/* TODO Add concurrency control by sequential lock on nexthop_mac. */
+	if (!map->stale)
+		ether_addr_copy(&map->ha, nexthop_mac);
+	else
+		/* 00:00:00:00:00:00 represents invalid MAC address. */
+		memset(nexthop_mac, 0, sizeof(*nexthop_mac));
+
+	*pcall_again = true;
+}
+
+static int
+gk_hold_arp_and_nd(struct gk_lpm *ltbl, unsigned int lcore_id)
+{
+	unsigned int i;
+	int ret;
+	struct gk_policy *policies = ltbl->policy_tbl;
+	for (i = 0; i < ltbl->num_policies; i++) {
+		struct ipip_tunnel_info *tunnel;
+
+		if (policies[i].action != GK_FWD_GT)
+			continue;
+
+		tunnel = &policies[i].grantor_tunnel;
+		if (tunnel->flow.proto == ETHER_TYPE_IPv4) {
+			struct in_addr *ipv4_dst;
+			if (policies[i].is_gt_in_back_net)
+				ipv4_dst = (struct in_addr *)
+					&tunnel->flow.f.v4.dst;
+			else
+				ipv4_dst = &policies[i].gw->ip.v4;
+
+			ret = hold_arp(gk_arp_and_nd_req_cb, (void *)
```

Comment:
Sorry, I may misunderstand your idea. (1) Do you mean that the ``struct ipip_tunnel_info`` only contains the network layer information, and the encapsulation function looks like ``encapsulate(struct rte_mbuf *pkt, uint8_t priority, struct ipip_tunnel_info *info, struct ether_hdr *eth)``. (2) Could you please point out why the MAC address is wrong? For each gateway and Grantor server in the back network, there is only one ``struct ether_hdr`` is maintained. If the destination MAC address is updated, it should be consistent. (3) I maintained the Ethernet cache in the gateway entry, i.e., the ``struct ether_hdr``. Is it in a wrong place?